video/decode/vd_lavc.c

```diff
@@ -465,101 +464,110 @@ static void select_and_set_hwdec(struct mp_filter *vd)
                     }
-                    MP_WARN(vd, "Using emulated hardware decoding API.\n");
-                }
-            } else if (!hwdec->copying) {
-                // Most likely METHOD_INTERNAL, which often use delay-loaded
-                // VO support as well.
-                if (ctx->hwdec_devs) {
-                    struct hwdec_imgfmt_request params = {
-                        .imgfmt = pixfmt2imgfmt(hwdec->pix_fmt),
-                        .probing = hwdec_auto,
-                    };
-                    hwdec_devices_request_for_img_fmt(
-                        ctx->hwdec_devs, &params);
-                }
-            }
 
-            ctx->use_hwdec = true;
-            ctx->hwdec = *hwdec;
-            break;
-        }
-
-        talloc_free(hwdecs);
+                    const struct hwcontext_fns *fns =
+                                hwdec_get_hwcontext_fns(hwdec->lavc_device);
+                    if (fns && fns->is_emulated && fns->is_emulated(ctx->hwdec_dev)) {
+                        if (hwdec_auto) {
+                            MP_VERBOSE(vd, "Not using emulated API.\n");
+                            av_buffer_unref(&ctx->hwdec_dev);
+                            continue;
+                        }
+                        MP_WARN(vd, "Using emulated hardware decoding API.\n");
+                    }
+                } else if (!hwdec->copying) {
+                    // Most likely METHOD_INTERNAL, which often use delay-loaded
+                    // VO support as well.
+                    if (ctx->hwdec_devs) {
+                        struct hwdec_imgfmt_request params = {
+                            .imgfmt = pixfmt2imgfmt(hwdec->pix_fmt),
+                            .probing = hwdec_auto,
+                        };
+                        hwdec_devices_request_for_img_fmt(
+                            ctx->hwdec_devs, &params);
+                    }
+                }
 
-        if (!ctx->use_hwdec) {
-            if (!hwdec_auto && !hwdec_name_supported)
-                MP_WARN(vd, "Unsupported hwdec: %s\n", ctx->opts->hwdec_api);
-            MP_VERBOSE(vd, "No hardware decoding available for this codec.\n");
+                ctx->use_hwdec = true;
+                ctx->hwdec = *hwdec;
+                break;
+            }
+            if (ctx->use_hwdec)
+                break;
+            else if (!hwdec_auto && !hwdec_name_supported)
+                MP_WARN(vd, "Unsupported hwdec: %.*s\n", BSTR_P(opt));
         }
     }
+    talloc_free(hwdecs);
+
 
     if (ctx->use_hwdec) {
         MP_VERBOSE(vd, "Trying hardware decoding via %s.\n", ctx->hwdec.name);
         if (strcmp(ctx->decoder, ctx->hwdec.codec->name) != 0)
             MP_VERBOSE(vd, "Using underlying hw-decoder '%s'\n",
                        ctx->hwdec.codec->name);
     } else {
+        if (hwdec_attempted)
+            MP_VERBOSE(vd, "No hardware decoding available for codec: %s\n",
+                       codec);
```

Comment:
is this really that useful when you have the "Looking at hwdec %s..." lines from above?